sys/net/dsm/dsm.c

```diff
@@ -61,7 +61,7 @@ dsm_state_t dsm_store(sock_dtls_t *sock, sock_dtls_session_t *session,
     }
 
     prev_state = session_slot->state;
-    if (session_slot->state != SESSION_STATE_ESTABLISHED) {
+    if (new_state > session_slot->state) {
```

Comment:
The reason I brought this in was to avoid that the state can be set back from `SESSION_STATE_HANDSHAKE` to `SESSION_STATE_SOCK_INIT`. This would occur on client side and mess up the gcoap flow in the handshake.
The only solution that comes to my mind right now would be another if nesting to prevent this exact case. Which doesn't seem to me like a nice solution.
Would there be something better to do this or is that solution ok?

Reply:
So, AFAIU this is checking that the state transition is legal.  Maybe the best would be to concentrate such logic in its own function, that way it could scale in case we need to add more states.

Reply:
Something like this?
```
bool _legal_state_change(dsm_state_t new_state, dsm_state_t old_state) {
    switch(old_state) {
        case SESSION_STATE_NONE:
            return true;
        case SESSION_STATE_SOCK_INIT:
            return ((new_state == SESSION_STATE_HANDSHAKE) || 
                    (new_state == SESSION_STATE_ESTABLISHED));
        case SESSION_STATE_HANDSHAKE:
            return new_state == SESSION_STATE_ESTABLISHED;
        case SESSION_STATE_ESTABLISHED:
            /* Fall through */
        default:
            return false;
    }
}
```

Reply:
maybe the function could be `_set_state` and actually apply the change if legal.

Reply:
Ok, done.